### PR TITLE
Product addtocart form validate

### PR DESCRIPTION
**Pull request by mehdichaouch**

### CODE REVIEW

**Comment by mehdichaouch** [comment]
Only last commit is about the branch feature.

File: app/design/frontend/base/default/template/oyst/oneclick.phtml [hunk 1]
```diff
@@ -23,49 +23,20 @@ $oneClickUrl = $this->getOneClickUrl();
...
-                            }
-                            cb(isErrorsInForm, data.url);
-                        }
-                    };
-                    xhr.send(form);
-                });
-            };
+            if ("function" === typeof oystOneClick ) {
+                oystOneClick(
+                    <?php echo $product->getId() ?>,
```
Comment (jney) [inline app/design/frontend/base/default/template/oyst/oneclick.phtml]: ce n'est pas un string ?

File: app/design/frontend/base/default/template/oyst/oneclick.phtml [hunk 2]
```diff
@@ -23,49 +23,20 @@ $oneClickUrl = $this->getOneClickUrl();
...
+                    "<?php echo $oneClickUrl ?>",
```
Comment (jney) [inline app/design/frontend/base/default/template/oyst/oneclick.phtml]: safer:

```php
<?php echo json_encode($oneClickUrl) ?>,
```

without quotes

File: app/design/frontend/base/default/template/oyst/oneclick.phtml [hunk 3]
```diff
@@ -23,49 +23,20 @@ $oneClickUrl = $this->getOneClickUrl();
...
+                    <?php echo $isProductAddtocartFormValidate ?>
```
Comment (jney) [inline app/design/frontend/base/default/template/oyst/oneclick.phtml]: same comment as above

File: js/oyst/oneclick.js [hunk 4]
```diff
@@ -11,10 +11,18 @@
...
     $("messages").update(html);
-    var url = location.href.split('#')[0]
-    location.replace(url + "#html-body");
+    var url = window.location.href.split("#")[0];
+    window.location.replace(url + "#html-body");
+}
+
+if (Validation) {
+    Validation.add('validate-oyst-url', 'Please enter a valid URL. Protocol is required (http:// or https://), finishing by "/"', function (v) {
+        v = (v || '').replace(/^\s+/, '').replace(/\s+$/, '');
```
Comment (jney) [inline js/oyst/oneclick.js]: isn't it `trim` function job?

File: js/oyst/oneclick.js [hunk 5]
```diff
@@ -11,10 +11,18 @@
...
...) || /^(https?):\/\/(((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:)*@)?(((\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5]))|((([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.)+(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.?)(:\d*)?)(\/((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)+(\/(([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)*)*)?)?(\?((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)|[\uE000-\uF8FF]|\/|\?)*)?(\#((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)|\/|\?)*)?$/i.test(v));
```
Comment (jney) [inline js/oyst/oneclick.js]: i don't get what are those regex doing, can you have an explicit comment?

File: app/design/frontend/base/default/template/oyst/oneclick.phtml (hunk 1 above)
Comment (mehdichaouch) [reply app/design/frontend/base/default/template/oyst/oneclick.phtml]: image
Je test la présente de la fonction

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 6]
```diff
@@ -11,7 +11,68 @@
...
+                url: oneClickUrl
+            };
+
+            var xhr = new XMLHttpRequest();
+            xhr.open(settings.method, settings.url, settings.async);
+            xhr.setRequestHeader('cache-control', settings.cacheControl);
+            xhr.onload = function () {
+                if (200 === xhr.status) {
+                    var data = JSON.parse(xhr.responseText);
+                    var isErrorsInForm = null;
```
Comment (jney) [inline skin/frontend/base/default/js/oyst/oneclick.js]: why not `false` by default?

File: js/oyst/oneclick.js (hunk 5 above)
Comment (mehdichaouch) [reply js/oyst/oneclick.js]: To validate that url end with "/".

File: skin/frontend/base/default/js/oyst/oneclick.js [hunk 7]
```diff
@@ -11,7 +11,68 @@
...
+ * @param oneClickUrl
+ * @param isProductAddtocartFormValidate
+ */
+function oystOneClick(productId, oneClickUrl, isProductAddtocartFormValidate) {
+    window.__OYST__ = window.__OYST__ || {};
+    window.__OYST__.getOneClickURL = function (cb, opts) {
+        opts = opts || {};
+
+        ready(function () {
+            var qty = $$("input[name='qty']")[0].value ? $$("input[name='qty']")[0].value : 1;
```
Comment (jney) [inline skin/frontend/base/default/js/oyst/oneclick.js]: shorter:
```js
var qty = $$("input[name='qty']")[0].value || 1
```

File: skin/frontend/base/default/js/oyst/oneclick.js (hunk 6 above)
Comment (mehdichaouch) [reply skin/frontend/base/default/js/oyst/oneclick.js]: null means they no form validation, true/false, is for form validation.

File: skin/frontend/base/default/js/oyst/oneclick.js (hunk 7 above)
Comment (jney) [reply skin/frontend/base/default/js/oyst/oneclick.js]: but be careful of js falsies

File: app/design/frontend/base/default/template/oyst/oneclick.phtml (hunk 2 above)
Comment (mehdichaouch) [reply app/design/frontend/base/default/template/oyst/oneclick.phtml]: Ok.

File: app/design/frontend/base/default/template/oyst/oneclick.phtml (hunk 3 above)
Comment (mehdichaouch) [reply app/design/frontend/base/default/template/oyst/oneclick.phtml]: Ok.

File: js/oyst/oneclick.js [hunk 8]
```diff
@@ -11,10 +11,18 @@
 /**
  * OneClick Customs JS
  */
+"use strict";
 
 function showMessage(txt, type) {
-    var html = '<ul class="messages"><li class="' + type + '-msg"><ul><li>' + txt + '</li></ul></li></ul>';
+    var html = "<ul class=\"messages\"><li class=\"" + type + "\"-msg\"><ul><li>" + txt + "</li></ul></li></ul>";
```
Comment (jney) [inline js/oyst/oneclick.js]: before the change, it looked clearer to me

Comment (mehdichaouch) [reply js/oyst/oneclick.js]: Yes, but with "use strict", I have to use " and not ' to surround code.

File: app/design/frontend/base/default/template/oyst/oneclick.phtml (hunk 1 above)
Comment (nicolaschenet) [reply app/design/frontend/base/default/template/oyst/oneclick.phtml]: Why not just `if(oystOneClick)` ?

File: js/oyst/oneclick.js (hunk 8 above)
Comment (nicolaschenet) [reply js/oyst/oneclick.js]: And you can't use template strings I guess ? 
```
var html = `
  <ul class="messages">
    <li class="${type}-msg">
      <ul>
        <li>${txt}</li>
      </ul>
    </li>
   </ul>
`
```

File: app/design/frontend/base/default/template/oyst/oneclick.phtml (hunk 1 above)
Comment (jney) [reply app/design/frontend/base/default/template/oyst/oneclick.phtml]: we are probably not talking about the same line.
i wondered why there are no quotes around `<?php echo $product->getId() ?>`.
Isn't it supposed to be a string?

File: js/oyst/oneclick.js (hunk 5 above)
Comment (nicolaschenet) [reply js/oyst/oneclick.js]: I guess you can do a simpler regexp...

Comment (jney) [reply js/oyst/oneclick.js]: what about `/\/$/.test(v)`?

Comment (cyrilhagege) [reply js/oyst/oneclick.js]: `str[str.length-1] === '/'` ?

Comment (jney) [reply js/oyst/oneclick.js]: or `v.endsWith('/')`

File: skin/frontend/base/default/js/oyst/oneclick.js (hunk 7 above)
Comment (nicolaschenet) [reply skin/frontend/base/default/js/oyst/oneclick.js]: @jney your code set qty to 1 if original qty is 0

File: app/design/frontend/base/default/template/oyst/oneclick.phtml (hunk 1 above)
Comment (mehdichaouch) [reply app/design/frontend/base/default/template/oyst/oneclick.phtml]: @jney of course product_reference is a string so I can surround it with quote.

Comment (mehdichaouch) [reply app/design/frontend/base/default/template/oyst/oneclick.phtml]: @nicolaschenet if I do it with `if(oystOneClick)`and it's not available I'll get an error `Uncaught ReferenceError: oystOneClick is not defined    at <anonymous>:1:1` is that a goo practice ?

File: js/oyst/oneclick.js (hunk 8 above)
Comment (jney) [reply js/oyst/oneclick.js]: weirdo
http://ecma262-5.com/ELS5_HTML.htm#Section_14.1

```
A Use Strict Directive is an ExpressionStatement in a Directive Prologue whose StringLiteral is either the exact character sequences "use strict" or 'use strict'. A Use Strict Directive may not contain an EscapeSequence or LineContinuation.
```

File: js/oyst/oneclick.js (hunk 5 above)
Comment (mehdichaouch) [reply js/oyst/oneclick.js]: Someone else has an other solution :D ?

File: js/oyst/oneclick.js (hunk 4 above)
Comment (mehdichaouch) [reply js/oyst/oneclick.js]: `validate-oyst-url ` validator is an adaptation of https://github.com/OpenMage/magento-mirror/blob/d48bebc211cc216aaf78bdf25d7f0b0143d6333b/js/prototype/validation.js#L576

File: skin/frontend/base/default/js/oyst/oneclick.js (hunk 6 above)
Comment (mehdichaouch) [reply skin/frontend/base/default/js/oyst/oneclick.js]: This is ok for you?

File: app/design/frontend/base/default/template/oyst/oneclick.phtml (hunk 3 above)
Comment (mehdichaouch) [reply app/design/frontend/base/default/template/oyst/oneclick.phtml]: Fixed

**Pull request merged**